Mathlib/ModelTheory/Equivalence.lean

```diff
@@ -34,31 +33,129 @@ variable {M : Type*} [Nonempty M] [L.Structure M] [M ⊨ T]
 
 namespace Theory
 
+/-- `φ ⟹[T] ψ` indicates that `φ` implies `ψ` in models of `T`. -/
+protected def Imp (T : L.Theory) (φ ψ : L.BoundedFormula α n) : Prop :=
+  T ⊨ᵇ φ.imp ψ
+
+@[inherit_doc FirstOrder.Language.Theory.Imp]
+scoped[FirstOrder] notation:51 φ:50 " ⟹[" T "] " ψ:51 => Language.Theory.Imp T φ ψ
+
+namespace Imp
+
+@[refl]
+protected theorem refl (φ : L.BoundedFormula α n) : φ ⟹[T] φ := fun _ _ _ => id
+
+instance : IsRefl (L.BoundedFormula α n) T.Imp := ⟨Imp.refl⟩
+
+@[trans]
+protected theorem trans {φ ψ θ : L.BoundedFormula α n} (h1 : φ ⟹[T] ψ) (h2 : ψ ⟹[T] θ) :
+    φ ⟹[T] θ := fun M v xs => (h2 M v xs) ∘ (h1 M v xs)
+
+instance : IsTrans (L.BoundedFormula α n) T.Imp := ⟨fun _ _ _ => Imp.trans⟩
+
+end Imp
+
+section Imp
+
+lemma bot_imp (φ : L.BoundedFormula α n) : ⊥ ⟹[T] φ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_bot, false_implies]
+
+lemma imp_top (φ : L.BoundedFormula α n) : φ ⟹[T] ⊤ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_top, implies_true]
+
+lemma imp_sup_left (φ ψ : L.BoundedFormula α n) : φ ⟹[T] φ ⊔ ψ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_sup]
+  exact Or.inl
+
+lemma imp_sup_right (φ ψ : L.BoundedFormula α n) : ψ ⟹[T] φ ⊔ ψ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_sup]
+  exact Or.inr
+
+lemma sup_imp {φ ψ θ : L.BoundedFormula α n} (h₁ : φ ⟹[T] θ) (h₂ : ψ ⟹[T] θ) :
+    φ ⊔ ψ ⟹[T] θ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_sup]
+  exact fun h => h.elim (h₁ M v xs) (h₂ M v xs)
+
+lemma sup_imp_iff {φ ψ θ : L.BoundedFormula α n} :
+    (φ ⊔ ψ ⟹[T] θ) ↔ (φ ⟹[T] θ) ∧ (ψ ⟹[T] θ) :=
+  ⟨fun h => ⟨(imp_sup_left _ _).trans h, (imp_sup_right _ _).trans h⟩,
+    fun ⟨h₁, h₂⟩ => sup_imp h₁ h₂⟩
+
+lemma inf_imp_left (φ ψ : L.BoundedFormula α n) : φ ⊓ ψ ⟹[T] φ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_inf]
+  exact And.left
+
+lemma inf_imp_right (φ ψ : L.BoundedFormula α n) : φ ⊓ ψ ⟹[T] ψ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_inf]
+  exact And.right
+
+lemma imp_inf {φ ψ θ : L.BoundedFormula α n} (h₁ : φ ⟹[T] ψ) (h₂ : φ ⟹[T] θ) :
+    φ ⟹[T] ψ ⊓ θ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_inf]
+  exact fun h => ⟨h₁ M v xs h, h₂ M v xs h⟩
+
+lemma imp_inf_iff {φ ψ θ : L.BoundedFormula α n} :
+    (φ ⟹[T] ψ ⊓ θ) ↔ (φ ⟹[T] ψ) ∧ (φ ⟹[T] θ) :=
+  ⟨fun h => ⟨h.trans (inf_imp_left _ _), h.trans (inf_imp_right _ _)⟩,
+    fun ⟨h₁, h₂⟩ => imp_inf h₁ h₂⟩
+
+lemma inf_not_imp_bot (φ : L.BoundedFormula α n) :
```

Comment:
This one and the next are `iff` in general, right? It seems more sensible to just give that version, then this one should be easily derivable from it

Reply:
I've now made both of these `iff`s and adjusted #16801 accordingly.